hca/dss/__init__.py

```diff
@@ -107,87 +109,182 @@ def download(self, bundle_uuid, replica, version="", dest_name="",
+                        if content_range_header is not None:
+                            cre = re.compile("bytes (\d+)-(\d+)")
+                            mo = cre.search(content_range_header)
+                            if mo is not None:
+                                server_start = int(mo.group(1))
+
+                        consume_bytes -= server_start
+                        assert consume_bytes >= 0
+                        if server_start > 0 and consume_bytes == 0:
+                            logger.info("%s", "File {}: Resuming at {}.".format(
+                                file_uuid, server_start))
+                        elif consume_bytes > 0:
+                            logger.info("%s", "File {}: Resuming at {}. Dropping {} bytes to match".format(
+                                file_uuid, server_start, consume_bytes))
+
+                            while consume_bytes > 0:
+                                bytes_to_read = min(consume_bytes, 1024*1024)
+                                content = response.iter_content(chunk_size=bytes_to_read)
+                                chunk = next(content)
+                                if chunk:
+                                    consume_bytes -= len(chunk)
+
+                        for chunk in response.iter_content(chunk_size=1024*1024):
+                            if chunk:
+                                fh.write(chunk)
+                                hasher.update(chunk)
+                                retries_left = min(retries_left + 1, num_retries)
+                                delay = max(delay / 2, min_delay_seconds)
+                        break
+                    finally:
+                        response.close()
+                except (ChunkedEncodingError, ConnectionError, ReadTimeout):
+                    if retries_left > 0:
+                        logger.info("%s", "File {}: GET FAILED. Attempting to resume.".format(
+                            file_uuid, dest_path))
+                        time.sleep(delay)
+                        delay *= 2
+                        retries_left -= 1
+                        continue
+                    raise
+
+        if hasher.hexdigest().lower() != file_sha256.lower():
+            os.remove(dest_path)
+            logger.error("%s", "File {}: GET FAILED. Checksum mismatch.".format(file_uuid))
+            raise ValueError("Expected sha256 {} Received sha256 {}".format(
+                file_sha256.lower(), hasher.hexdigest().lower()))
+        else:
+            logger.info("%s", "File {}: GET SUCCEEDED. Stored at {}.".format(file_uuid, dest_path))
+
+    @classmethod
+    def _file_path(cls, checksum):
+        """
+        returns a file's relative local path based on the nesting parameters and the files hash
+        :param checksum: a string checksum
+        :return: relative Path object
+        """
+        assert(sum(cls.FILE_NESTING_PARAMETER) <= len(checksum))
```

Comment:
This assert will fail if the FILE_NESTING_PARAMETER is [15, 18] and the checksum length is 32, even though that seems perfectly valid to me.

Reply:
Would you need to worry about things like `[2, 2, 2]`?  I don't imagine this needs to be configured ever, and I'm kind of arguing for not needing it at all above, but I was imagining something more like:
```
i = sorted(cls.FILE_NESTING_PARAMETER)
assert((i[-1] <= len(checksum)) and i[0] >= 0 and len(set(i)) == len(i))
```
Please correct me if I'm misunderstanding.

Reply:
`[2, 2, 2]` is fine. And yes, I think you are misunderstanding. AFAICT, @jessebrennan's implementation is correct and straight-forward.